lib/programmemory.cpp

```diff
@@ -229,12 +230,12 @@ static void fillProgramMemoryFromAssignments(ProgramMemory& pm, const Token* tok
                 pm.setUnknown(vartok->varId());
         }
 
-        if (tok2->str() == "{") {
+        if (tok2->str() == MatchCompiler::makeConstString("{")) {
```

Comment:
This is the only place in cppcheck that uses `MatchCompiler::makeConstString()` directly. Would using `Token::simpleMatch(tok2, "{")` have the same effect?

Also, are you saying that if a `str::string` is compared with a `const char *` that is known at compile-time, the compiler still generates a call to `strlen()`? I would have expected something smarter from the compiler, but perhaps I'm naive... If that's the case perhaps it would be beneficial to upgrade `matchcompiler.py` to replace **all** such calls with `MatchCompiler::makeConstString()`?

Reply:
> This is the only place in cppcheck that uses `MatchCompiler::makeConstString()` directly. Would using `Token::simpleMatch(tok2, "{")` have the same effect?

Argh - yeah my IDE keeps mixing up the normal sources and the matchcompiler ones because of `#line` - so I need to adjust that.

> Also, are you saying that if a `str::string` is compared with a `const char *` that is known at compile-time, the compiler still generates a call to `strlen()`? I would have expected something smarter from the compiler, but perhaps I'm naive... If that's the case perhaps it would be beneficial to upgrade `matchcompiler.py` to replace **all** such calls with `MatchCompiler::makeConstString()`?

Yes, it does. Same if you pass a literal as a `std::string` parameter it will always generate a temporary object of that. Quite annoying - I assume `constexpr` might make things better but I have not much experience with that in terms of function calls.

That might be a good idea to extend the MatchCompiler in that way instead. In my current test it was just a tiny percentage - but there might be cases where it is more.

Reply:
The compiler does get rid of `strlen` calls when the string is constant(see [here](https://godbolt.org/z/6jYfYK)). However, it doesn't seem to inline `std::string::compare` function, unless you use libc++(see [here](https://godbolt.org/z/f3aqM9)). These kind of optimizations seem like they would be better upstreamed into libstdc++.

Reply:
@pfultz2 `libstdc++` seems to be fine with GCC. I am always using Clang with libstdc++ (the default). I could also profile it with GCC for comparison.

Reply:
It looks like libc++ with -O3 actually produces worse code than -O2 for the comparison - I guess that might be worth a bug report. Using libstdc++ there is no difference between the optimizations.

libc++ also produces worse code than libstdc++.